Sources/SwiftParser/Diagnostics/ParserDiagnosticMessages.swift

```diff
@@ -131,79 +132,184 @@ public struct InvalidIdentifierError: ParserError {
+    }
+    return nil
+  }
+
+  /// If applicable, returns a string that describes the node in which the missing nodes are expected.
+  private var parentContextClause: String? {
+    enum ContextType {
+      case toStart
+      case `in`
+      case toEnd
+    }
+    var nodePosition = ContextType.in
+
+    if let missingToken = missingNodes.first?.as(TokenSyntax.self), let commonParent = commonParent {
       switch missingToken.tokenKind {
-      case .leftAngle, .leftBrace, .leftParen, .leftSquareBracket:
-        if parent.children(viewMode: .fixedUp).first?.as(TokenSyntax.self) == missingToken {
-          message += " to start \(parentTypeName)"
+      case .leftBrace, .leftAngle, .leftParen, .leftSquareBracket:
+        if commonParent.children(viewMode: .fixedUp).first?.as(TokenSyntax.self) == missingToken {
+          nodePosition = .toStart
         }
-      case .rightAngle, .rightBrace, .rightParen, .rightSquareBracket:
-        if parent.children(viewMode: .fixedUp).last?.as(TokenSyntax.self) == missingToken {
-          message += " to end \(parentTypeName)"
+      default:
+        break
+      }
+    }
+    if let missingToken = missingNodes.last?.as(TokenSyntax.self), let commonParent = commonParent {
+      switch missingToken.tokenKind {
+      case .rightBrace, .rightAngle, .rightParen, .rightSquareBracket:
+        if commonParent.children(viewMode: .fixedUp).last?.as(TokenSyntax.self) == missingToken {
+          if nodePosition == .toStart {
+            // If the missing tokens encomposs both the start and end, emit an 'in' context clause
+            nodePosition = .in
+          } else {
+            nodePosition = .toEnd
+          }
         }
       default:
-        message += " in \(parentTypeName)"
+        break
       }
     }
-    return message
+    if let commonParent = commonParent, let parentTypeName = commonParent.nodeTypeNameForDiagnostics(allowSourceFile: false) {
+      switch nodePosition {
+      case .toStart:
+        return "to start \(parentTypeName)"
+      case .in:
+        return "in \(parentTypeName)"
+      case .toEnd:
+        return "to end \(parentTypeName)"
+      }
+    }
+    return nil
   }
 
-  public var handledNodes: [Syntax] {
-    if let previous = missingToken.previousToken(viewMode: .all), previous.parent!.is(UnexpectedNodesSyntax.self) {
-      return [Syntax(previous.parent!)]
+  public var message: String {
+    var message = "Expected \(self.missingNodesDescription)"
+    if let afterClause = afterClause {
+      message += " \(afterClause)"
+    }
+    if let parentContextClause = parentContextClause {
+      message += " \(parentContextClause)"
     }
-    return []
+    return message
+  }
+}
+
+public struct MissingAttributeArgument: ParserError {
+  /// The name of the attribute that's missing the argument, without `@`.
+  public let attributeName: TokenSyntax
+
+  public var message: String {
+    return "Expected argument for '@\(attributeName)' attribute"
```

Comment:
`argument` or `arguments`?

Reply:
I think based on the way we currently model attributes we don’t really know whether a particular attribute requires a single argument or multiple… Might be something to tackle in a follow-up PR.

Reply:
`argument(s)` 😅